[FLINK-13513][ml] Add Mapper and related classes for later algorithms

xuyang1706 opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Travis CI to do that following [this guide](https://flink.apache.org/contributing/contribute-code.html#open-a-pull-request).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change

Add  the FlatMapper and related classes  for later algorithm implementations. Define the abstract classes for the single-thread executor for Row type data.


## Brief change log

  - *Add the definition of abstract class FlatMapper*
  - *Add interface MapOpInterface for  the map operation of Row type data.*
  - *Add abstract class FlatModelMapper, Mapper extends FlatMapper*
  - *Add abstract class ModelMapper extends FlatModelMapper*


## Verifying this change
This change added tests and can be verified as follows:

- run test case pass

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

weibo611 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java:
```diff
@@ -0,0 +1,122 @@
+	}
+
+	/**
+	 * Get the table schema(includs column names and types) of the calculation result.
```
"includs" is a spell error

xuyang1706 replied:
thx, it has been fixed

weibo611 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java:
```diff
@@ -0,0 +1,50 @@
+
+/**
+ * Abstract class for mappers with model.
+ * ModelMapper transform one Row type data into one Row type result data.
```
"transform" should be "transforms"

xuyang1706 replied:
thx, it has been fixed

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java:
```diff
@@ -0,0 +1,129 @@
+	 * The instance can not deal with real data, but it could be used to get the output result schema.
+	 *
+	 * @param flatModelMapperClassName Name of the FlatModelMapper class
+	 * @param modelScheme              The schema of input Table type model.
```
it is not immediately clear to me what's the difference between `modelScheme` &  `dataSchema`. 
Also why is modelScheme of the type `TableSchema` ?

xuyang1706 replied:
Thanks, the model is the machine learning model that use the Table as its representation (serialized to Table from the memory or deserialized from Table to memory). Thus, the model need `modelSchema` and the predict data need `dataSchema`.

walterddr replied:
ah. I see now. I think the prhase "the schema of the input Table" part confused me.
I would probably rephrase it to
```
@param modelScheme    The model scheme represented in Table format
```

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java:
```diff
@@ -0,0 +1,129 @@
+
+/**
+ * Abstract class for flatMappers with model.
+ * FlatModelMapper transform one Row type data into zero, one, or more Row type result data.
```
IMO, This is duplicate Javadoc similar to `FlatMapper`. Can we instead explain how the "model" part works. Basic `FlapMapper` already explains these.

xuyang1706 replied:
Thanks, removed.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOpInterface.java:
```diff
@@ -0,0 +1,38 @@
+/**
+ * Interface for the map operation of Row type data.
+ */
+public interface MapOpInterface {
```
should this file be named just `MapOp`. There's already a `MatVecOp` class thought we would like to keep the naming conversion as well?

xuyang1706 replied:
Thanks for your advice, I renamed this interface to `MapOperable`.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java:
```diff
@@ -0,0 +1,122 @@
+/**
+ * Abstract class for flatMappers.
+ * FlatMapper transform one Row type data into zero, one, or more Row type result data.
+ * Operations that produce multiple strictly one Row type result data per Row type data
```
Thanks, FlatMapper can not support multiple rows as its inputs and we have been removed the `Iterable<Row>` in this interface

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java:
```diff
@@ -0,0 +1,122 @@
+	 * @param output the output collector
+	 * @throws Exception if {@link #flatMap(Row, Collector)} throws exception.
+	 */
+	public void flatMap(Iterable <Row> rows, Collector<Row> output) throws Exception {
```
Thanks, removed

xuyang1706 commented:
> Thanks for the contribution @xuyang1706 . I took a quick look and I think there are some confusions I have with the Javadocs. Can you please take a look and let me know if any of my questions make sense?

@walterddr , thanks for your comments and suggestion. I have refined the JavaDoc and add more descriptions, removed unnecessary method and renamed the interface name.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java:
```diff
@@ -0,0 +1,138 @@
+	 *
+	 * @param flatModelMapperClassName Name of the FlatModelMapper class
+	 * @param modelScheme              The schema of input Table type model.
+	 * @param dataSchema               The schema of input Table type data.
```
```
@param dataSchema    The schema of the input data represented in Table format
```

xuyang1706 replied:
Thanks, changed.

ex00 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java:
```diff
@@ -0,0 +1,49 @@
+	 * @throws Exception if {@link Collector#collect(Object)} throws exception.
+	 */
+	@Override
+	public void flatMap(Row row, Collector<Row> output) throws Exception {
```
it is same as in `ModelMapper`
I suppose this logic can be move to `FlatMapper` class directly, because in this version of code base all implementation this method is equal.

xuyang1706 replied:
Thanks for your advice. For One of the extended class `FlatModelMapper` not support the map() function, we can not override its flatMap() function. Thus, I can not move the function to  FlatMapper class.

xuyang1706 replied:
Yes, there are other good ways to implement. I'd like to choose current one. Thanks for your helpful discussion and advice!

xuyang1706 commented:
> thanks for the update @xuyang1706 and sorry for the late reply. Overall I think the change looks good to me. @becketqin do you have any additional comments?

Thanks @walterddr , your suggested descriptions are much better, I have rephrased them.

xuyang1706 commented:
> Hi @xuyang1706
> Thanks for you PR. I have left only one minor comment, please look it if you have time.
> In general this changes LGTM.
> Thanks!

Thanks @ex00 for your helpful comments and suggestion.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java:
```diff
@@ -0,0 +1,108 @@
+	 * @return new instance of given FlatMapper class
+	 * @throws Exception if flatMapperClass is not the class of {@link FlatMapper}
+	 */
+	public static FlatMapper of(
```
Thanks, changed the static method.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatModelMapper.java:
```diff
@@ -0,0 +1,138 @@
+	 * @return The object of {@link FlatModelMapper}
+	 * @throws Exception if flatModelMapperClass is not the class of {@link FlatModelMapper}
+	 */
+	public static FlatModelMapper of(
```
Thanks, changed the static method.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/MapOperable.java:
```diff
@@ -0,0 +1,40 @@
+ *
+ * <p>MapOperable maps a row to a new row.
+ */
+public interface MapOperable {
```
Thanks, refactored the definition of Mapper.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java:
```diff
@@ -0,0 +1,49 @@
+/**
+ * Abstract class for mappers.
+ */
+public abstract class Mapper extends FlatMapper implements MapOperable {
```
Thanks, refactored the definition of Mapper.

xuyang1706 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/FlatMapper.java:
```diff
@@ -0,0 +1,108 @@
+		Params params) throws Exception {
+
+		if (FlatMapper.class.isAssignableFrom(flatMapperClass)) {
+			return (FlatMapper) flatMapperClass.getConstructor(TableSchema.class, Params.class)
```
Thanks, changed.

xuyang1706 commented:
> @xuyang1706 Thanks for the patch. The structure and semantic in the patch seems a little confusing to me. I created a PR against your PR with my suggestions. The PR might be a little messy at this point. But should be clean once you rebase the `flatmapper` branch on master.
> 
> Please let me know what do you think. Thanks.

Thanks for your comments and discussion, @becketqin. We refactored the codes, please look it if you have time.  Thanks, -Xu

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java:
```diff
@@ -0,0 +1,123 @@
+		 * Load model data from data rows.
+		 */
+		Rows
+	}
```
It is probably worth separating out the model source so it can be reused by other classes. Also it seems indication that we may need two ModelSource implementations. I have created a PR against your PR branch with my suggestions.

xuyang1706 replied:
Yes, it is better to separate out the model source.

becketqin commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java:
```diff
@@ -0,0 +1,123 @@
+		return TableSchema.builder().fields(this.modelFieldNames, this.modelFieldTypes).build();
+	}
+
+	public final void setModelBroadcastName(String modelBroadcastName) {
```
Do we know the ModelSource at construction time? If so, the model source can be passed to the constructor to make the class immutable.

xuyang1706 replied:
Thanks, the model source is passed to the constructor.

becketqin commented:
Also, just curious what about the `FlatMapper` classes? Do we still want them?

xuyang1706 commented:
> Also, just curious what about the `FlatMapper` classes? Do we still want them?

Thanks for your advice, @becketqin. The `FlatMapper` class is removed from this PR, and will be submitted later with its related functionalities.

walterddr commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/ModelMapper.java:
```diff
@@ -0,0 +1,78 @@
+	@Override
+	public final void open(Configuration parameters) {
+		Preconditions.checkState(this.modelSource != null, "model source not set.");
+		List<Row> rows = this.modelSource.getModelRows(getRuntimeContext());
```
Shouldn't the parameters be part of the interface signature?

qiuxiafei replied:
Actually it shouldn't, we're working on an update which decouples ModelMapper and Mapper  from flink's `MapFunction`.

xuyang1706 replied:
We have updated the code.

walterddr replied:
I see. however at this current stage it seemed weird that the open method does not use `Configuration` passed in, instead using the more generic `getRuntimeContext`. 
I am find with current impl actually. thanks for the explanation.

bowenli86 commented on flink-ml-parent/flink-ml-lib/src/main/java/org/apache/flink/ml/common/mapper/Mapper.java:
```diff
@@ -0,0 +1,75 @@
+	/**
+	 * schema of the input.
+	 */
+	private final String[] dataFieldNames;
```
Why not just keep the dataSchema passed in from constructor? It seems a bit unintuitive to me

qiuxiafei replied:
It not serializable while the mapper class will be serialized to TM for execution.

xuyang1706 commented:
> Hi @xuyang1706 @becketqin Thanks for the discussion and sorry for the late response. This PR currently seems to be far away from the original intent of the [JIRA](https://issues.apache.org/jira/browse/FLINK-13513). I would suggest please change the title as well as the content for the commit message to make it more clear that we only support `ModelMapper` and not FlatMapper.

Thanks for your suggestion, @walterddr . I have changed the title.

becketqin commented:
@xuyang1706 I created a PR with some minor suggestion. Most of them are just java docs. Can you check and see if you have any concerns. I'll merge the code if you are OK with the changes. Thanks.

Thanks for the review @walterddr @bowenli86 @ex00 @weibo611

xuyang1706 commented:
> @xuyang1706 I created a PR with some minor suggestion. Most of them are just java docs. Can you check and see if you have any concerns. I'll merge the code if you are OK with the changes. Thanks.
> 
> Thanks for the review @walterddr @bowenli86 @ex00 @weibo611

Thanks @becketqin , your changes are good to me, please merge this PR.

walterddr commented:
close via [005bda9be](https://github.com/apache/flink/commit/005bda9be)
